### ISSUE TITLE
[FR] pkg_resources.parse_requirements: supporting nested requirements.txt

### ISSUE THREAD

**Issue by jamesbraza** [pypa/setuptools#3465]
### What's the problem this feature will solve?

I parse `requirements.txt` with `pkg_resources.parse_requirements`, and this has worked great in the normal case.

It's also possible to have nested `requirements.txt`:

```txt
-r another-requirements.txt
package==1.0.0
```

Currently, `pkg_resources.parse_requirements` doesn't support this, it throws an `InvalidRequirement` exception:

```txt
  File "/path/to/venv/lib/python3.10/site-packages/pkg_resources/__init__.py", line 3084, in parse_requirements
    yield Requirement(line)
  File "/path/to/venv/lib/python3.10/site-packages/pkg_resources/__init__.py", line 3094, in __init__
    super(Requirement, self).__init__(requirement_string)
  File "/path/to/venv/lib/python3.10/site-packages/pkg_resources/_vendor/packaging/requirements.py", line 100, in __init__
    raise InvalidRequirement(
pkg_resources.extern.packaging.requirements.InvalidRequirement: Parse error at "'-r requi'": Expected W:(abcd...)
```

### Describe the solution you'd like

It would be nice to have `pkg_resources.parse_requirements` support reading in nested requirements files.

This would probably entail `pkg_resources.parse_requirements` recursively opening a nested requirements file, and parsing it, while checking if a line starts with `"-r"` to recurse.

### Alternative Solutions

Here is an example showing how to accommodate one-layer of nesting.  A better solution would be recursively implemented, and have a max depth argument

```python
with open("requirements.txt", encoding="utf-8") as outer_f:
    requirements: list[str] = []
    for req in outer_f.readlines():
        if req.startswith("-r"):
            with open(req.split("-r")[1].strip(), encoding="utf-8") as inner_f:
                requirements += inner_f.readlines()
        else:
            requirements += [req]
    install_requires = [
        str(requirement)
        for requirement in pkg_resources.parse_requirements(requirements)
    ]
```

### Code of Conduct

- [X] I agree to follow the PSF Code of Conduct

**Comment by abravalheri** [comment]
Hi @jamesbraza, sorry for getting back to you this late.

The main motivation behind `pkg_resources.parse_requirements` is to parse a string where each line corresponds to a dependency specification. This is not compatible with `requirements.txt` files, since this kind of files may contain `pip` specific logic (for example the `-e` flag), and therefore do not fit in the same mindset.

I personally don't think this kind of feature would fit nicely within the `pkg_resources` or `setuptools` packages (it seems to be a bit outside of the objectives of the project, considering the previous issues/discussions about `requirements.txt`).

Have you considered some other alternatives, like using [`pip-tools`](https://github.com/jazzband/pip-tools/) to expand the file into a proper flat list of requirements?

**Comment by jamesbraza** [comment]
Hello @abravalheri thank you for the response!

Yes I am aware of `pip-tools`, unfortunately `pip-tools` currently doesn't support what I am looking to do.  Hence why I am DIY'ing stuff.  I was hoping to not re-invent a requirements parser, and came across `pkg_resources.parse_requirements` as a viable option.

Currently I do employ a DIY solution, so I am not blocked by this, it would just be nice to use a 3rd party library.

> This is not compatible with `requirements.txt` files, since this kind of files may contain `pip` specific syntax (for example the `-e` flag), and therefore do not fit in the same mindset.

I was not aware of this, thank you for pointing out!

It sounds then like `setuptools` doesn't support this, and doesn't plan on it, so feel free to close this ticket out.

**Comment by abravalheri** [comment]
Let's close this issue for now. The main idea is that in the future `pkg_resources` will be replaced by more optimised tools like `importlib.metadata` and `importlib.resources`. Because of that, it is very unlikely that we want to augment the scope of `pkg_resources`.
